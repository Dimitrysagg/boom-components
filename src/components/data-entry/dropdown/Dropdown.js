import React from 'react'
import PropTypes from 'prop-types'

const Dropdown = ({ ...props }) => <div />

Dropdown.propTypes = {
  /** whether dropdown menu is disabled */
  disabled: PropTypes.bool,
  /** Set the container of the dropdown menu. The default is to create a div element in body, you can reset it to the scrolling area and make a relative reposition. */
  getPopupContainer: PropTypes.func,
  /** the dropdown content */
  content: PropTypes.node,
  /** placement of content */
  position: PropTypes.oneOf(['bottomCenter', 'bottomLeft', 'bottomRight', 'topCenter', 'topLeft', 'topRight']),
  /** trigger which executes the dropdown action (hover doesn't work on mobile devices) */
  trigger: PropTypes.arrayOf(PropTypes.oneOf(['click', 'contextMenu', 'hover'])),
  /** callback when content show or hide */
  onVisibleChange: PropTypes.func
}

Dropdown.defaultProps = {
  disabled: false,
  getPopupContainer: () => document.body,
  position: 'bottomLeft',
  trigger: ['hover']
}

export default Dropdown
