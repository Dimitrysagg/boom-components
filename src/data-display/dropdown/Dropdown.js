/* eslint-disable indent */
import React, { Fragment, Component, Children, cloneElement } from "react";
import PropTypes from "prop-types";
import RcDropdown from "rc-dropdown";
import cx from "classnames";
import { noop } from "../../helpers";

import { DropdownStyle } from "./styles";

class Dropdown extends Component {
  getTranstionName = () => {
    const { placement = "", transitionName } = this.props;
    if (transitionName !== undefined) {
      return transitionName;
    }
    if (placement.indexOf("top") >= 0) {
      return "slide-down";
    }
    return "slide-up";
  };

  renderOverlay = () => {
    const { overlay } = this.props;
    let overlayNode;
    if (typeof overlay === "function") {
      overlayNode = overlay();
    } else {
      overlayNode = overlay;
    }
    overlayNode = Children.only(overlayNode);
    const overlayProps = overlayNode.props;
    const { selectable = false, focusable = true } = overlayProps;
    const fixedModeOverlay =
      typeof overlayNode.type === "string"
        ? overlay
        : cloneElement(overlayNode, {
            mode: "vertical",
            selectable,
            focusable
          });
    return fixedModeOverlay;
  };

  render() {
    const { children, trigger, disabled, prefixCls, showArrow } = this.props;
    const child = Children.only(children);
    const dropdownTrigger = cloneElement(child, {
      className: cx(child.props.className)
    });

    const triggerActions = disabled ? [] : trigger;

    let alignPoint;
    if (triggerActions && triggerActions.indexOf("contextMenu") !== -1) {
      alignPoint = true;
    }

    const prefixedCls = `${prefixCls}-dropdown`;

    return (
      <Fragment>
        <DropdownStyle showArrow={showArrow} />
        <RcDropdown
          alignPoint={alignPoint}
          {...this.props}
          transitionName={this.getTranstionName()}
          prefixCls={prefixedCls}
          trigger={triggerActions}
          overlay={() => this.renderOverlay(prefixedCls)}
        >
          {dropdownTrigger}
        </RcDropdown>
      </Fragment>
    );
  }
}

Dropdown.propTypes = {
  /** Whether or not the showing of contents inside dropdown is disabled */
  disabled: PropTypes.bool,
  /** Delay when hovering trigger element before showing dropdown contents */
  mouseEnterDelay: PropTypes.number,
  /** Delay when hovering trigger element before hidding dropdown contents */
  mouseLeaveDelay: PropTypes.number,
  /** function called when changing visibility of dropdown. Takes visibile as an argument */
  onVisibleChange: PropTypes.func,
  /** Placement of the dropdown relative to the trigger element */
  placement: PropTypes.oneOf([
    "topLeft",
    "topCenter",
    "topRight",
    "bottomLeft",
    "bottomCenter",
    "bottomRight"
  ]),
  /** Css classes for the dropdown will be prefixed with it */
  prefixCls: PropTypes.string,
  /** whether or not show an arrow when dropdown is visible */
  showArrow: PropTypes.bool,
  /** Action which triggers the display of dropdown. */
  trigger: PropTypes.oneOf(["click", "hover", "contextMenu"])
};

Dropdown.defaultProps = {
  disabled: false,
  mouseEnterDelay: 0.15,
  mouseLeaveDelay: 0.1,
  onVisibleChange: noop,
  placement: "bottomCenter",
  prefixCls: "alb",
  showArrow: false,
  trigger: "hover"
};

export default Dropdown;
