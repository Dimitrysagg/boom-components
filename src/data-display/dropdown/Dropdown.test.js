import React from "react";
import { shallow } from "enzyme";
import Dropdown from "./Dropdown";

describe("Collapse", () => {
  describe("Snapshot", () => {
    it("should render a collapse", () => {
      const menu = (
        <ul>
          <li>1st Menu Item</li>
          <li>2nd Menu Item</li>
          <li>3rd Menu Item</li>
        </ul>
      );
      const wrapper = shallow(
        <Dropdown menu={menu}>
          <a>Hover me!</a>
        </Dropdown>
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
});
