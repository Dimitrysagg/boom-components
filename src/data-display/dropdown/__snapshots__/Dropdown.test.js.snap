// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse Snapshot should render a collapse 1`] = `
<Fragment>
  <GlobalStyleComponent
    showArrow={false}
  />
  <Dropdown
    defaultVisible={false}
    disabled={false}
    menu={
      <ul>
        <li>
          1st Menu Item
        </li>
        <li>
          2nd Menu Item
        </li>
        <li>
          3rd Menu Item
        </li>
      </ul>
    }
    mouseEnterDelay={0.15}
    mouseLeaveDelay={0.1}
    onVisibleChange={[Function]}
    overlay={[Function]}
    overlayClassName=""
    overlayStyle={Object {}}
    placement="bottomCenter"
    prefixCls="alb-dropdown"
    showAction={Array []}
    showArrow={false}
    transitionName="slide-up"
    trigger="hover"
  >
    <a
      className=""
    >
      Hover me!
    </a>
  </Dropdown>
</Fragment>
`;
